odbc/src/api/functions.rs

```diff
@@ -1503,66 +1501,85 @@ pub unsafe extern "C" fn SQLGetConnectAttr(
     value_ptr: Pointer,
     buffer_length: Integer,
     string_length_ptr: *mut Integer,
 ) -> SqlReturn {
     panic_safe_exec!(
         || {
-            let mut err = None;
             let conn_handle = MongoHandleRef::from(connection_handle);
 
-            // This scope is introduced to make the RWLock Guard expire before we write
-            // any error values via add_diag_info as RWLock::write is not reentrant on
-            // all operating systems, and the docs say it can panic.
-            let sql_return = {
-                let conn = must_be_valid!((*conn_handle).as_connection());
-                let attributes = &conn.attributes.read().unwrap();
-
-                match attribute {
-                    ConnectionAttribute::CurrentCatalog => {
-                        let current_catalog = attributes.current_catalog.as_deref();
-                        match current_catalog {
-                            None => SqlReturn::NO_DATA,
-                            Some(cc) => i32_len::set_output_wstring(
-                                cc,
-                                value_ptr as *mut WChar,
-                                buffer_length as usize,
-                                string_length_ptr,
-                            ),
-                        }
-                    }
-                    ConnectionAttribute::LoginTimeout => {
-                        let login_timeout = attributes.login_timeout.unwrap_or(0);
-                        i32_len::set_output_fixed_data(&login_timeout, value_ptr, string_length_ptr)
-                    }
-                    ConnectionAttribute::ConnectionTimeout => {
-                        let connection_timeout = attributes.connection_timeout.unwrap_or(0);
-                        i32_len::set_output_fixed_data(
-                            &connection_timeout,
-                            value_ptr,
-                            string_length_ptr,
-                        )
-                    }
-                    _ => {
-                        err = Some(ODBCError::UnsupportedConnectionAttribute(
-                            connection_attribute_to_string(attribute),
-                        ));
-                        SqlReturn::ERROR
-                    }
+            match FromPrimitive::from_i32(attribute) {
+                Some(valid_attr) => sql_get_attrw_helper(
+                    conn_handle,
+                    valid_attr,
+                    value_ptr,
+                    buffer_length,
+                    string_length_ptr,
+                ),
+                None => {
+                    conn_handle.add_diag_info(ODBCError::InvalidAttrIdentifier(attribute));
+                    SqlReturn::ERROR
                 }
-            };
-
-            if let Some(error) = err {
-                conn_handle.add_diag_info(error);
             }
-            sql_return
         },
         connection_handle
     )
 }
 
+unsafe fn sql_get_attrw_helper(
```

Comment:
The reasons I added the helpers is actually a lot more trivial than what you're thinking!

One of the reasons they are specific to the unicode functions is because those are the only ones we have implemented and, like you pointed out, I figured we wouldn't be implementing the ANSI versions any time soon. Even when we _do_ implement the ANSI versions, we can always refactor at that time. I figured there was no need to over-abstract the code right now for functionality we don't plan on implementing.

The other main motivation behind moving the actual function logic into helpers was purely for readability purposes. If this code was left in the actual public API functions (the `SQL_XXX` functions), we'd have code so deeply nested, it would be indented between 7 and 12 times, causing line wraps. So ultimately, these helper functions are just helping keep the code a little more organized and readable.